Objects/bytesobject.c

```diff
@@ -2810,82 +2811,44 @@ _PyBytes_FromBuffer(PyObject *x)
 }
 
 static PyObject*
-_PyBytes_FromList(PyObject *x)
+_PyBytes_FromSequence(PyObject *x)
 {
-    Py_ssize_t i, size = PyList_GET_SIZE(x);
-    Py_ssize_t value;
-    char *str;
-    PyObject *item;
-    _PyBytesWriter writer;
-
-    _PyBytesWriter_Init(&writer);
-    str = _PyBytesWriter_Alloc(&writer, size);
-    if (str == NULL)
+    Py_ssize_t size = PySequence_Fast_GET_SIZE(x);
+    PyObject *bytes = _PyBytes_FromSize(size, 0);
+    if (bytes == NULL) {
         return NULL;
-    writer.overallocate = 1;
-    size = writer.allocated;
-
-    for (i = 0; i < PyList_GET_SIZE(x); i++) {
-        item = PyList_GET_ITEM(x, i);
-        Py_INCREF(item);
-        value = PyNumber_AsSsize_t(item, NULL);
-        Py_DECREF(item);
-        if (value == -1 && PyErr_Occurred())
-            goto error;
-
-        if (value < 0 || value >= 256) {
-            PyErr_SetString(PyExc_ValueError,
-                            "bytes must be in range(0, 256)");
-            goto error;
-        }
-
-        if (i >= size) {
-            str = _PyBytesWriter_Resize(&writer, str, size+1);
-            if (str == NULL)
-                return NULL;
-            size = writer.allocated;
-        }
-        *str++ = (char) value;
     }
-    return _PyBytesWriter_Finish(&writer, str);
-
-  error:
-    _PyBytesWriter_Dealloc(&writer);
-    return NULL;
-}
-
-static PyObject*
-_PyBytes_FromTuple(PyObject *x)
-{
-    PyObject *bytes;
-    Py_ssize_t i, size = PyTuple_GET_SIZE(x);
-    Py_ssize_t value;
-    char *str;
-    PyObject *item;
-
-    bytes = PyBytes_FromStringAndSize(NULL, size);
-    if (bytes == NULL)
-        return NULL;
-    str = ((PyBytesObject *)bytes)->ob_sval;
-
-    for (i = 0; i < size; i++) {
-        item = PyTuple_GET_ITEM(x, i);
-        value = PyNumber_AsSsize_t(item, NULL);
-        if (value == -1 && PyErr_Occurred())
+    char *str = PyBytes_AS_STRING(bytes);
+    PyObject *const *items = PySequence_Fast_ITEMS(x);
```

Comment:
If we're going for performance, then we can do even better. `PySequence_Fast_ITEMS` will call `PyList_Check`, but we already know that it's exactly a list or tuple here.

Reply:
But all we know is that it is either a list or a tuple, but we still don't know which of the two it is, so a `PyList_Check` call is still in order.

Reply:
Right, but `PyList_Check` is extra work because that will check for a subclass. It's not really going to be noticable, but it's something to think about :)

Reply:
Alternatively, you can just duplicate the functions as we did previously. This is not really an issue IMO, and we couold use the fact that tuples are immutables for instance to avoid INCREF/DECREF values.